runtime/vm/ContinuationHelpers.cpp

```diff
@@ -336,4 +337,55 @@ walkAllStackFrames(J9VMThread *currentThread, J9StackWalkState *walkState)
 									(void*)walkState);
 	return rc;
 }
+
+void
+acquireVThreadInspector(J9VMThread *currentThread, jobject thread)
+{
+	J9JavaVM *vm = currentThread->javaVM;
+	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+	MM_ObjectAccessBarrierAPI objectAccessBarrier = MM_ObjectAccessBarrierAPI(currentThread);
+	j9object_t threadObj = J9_JNI_UNWRAP_REFERENCE(thread);
+	I_64 vthreadInspectorCount;
+retry:
+	vthreadInspectorCount = J9OBJECT_I64_LOAD(currentThread, threadObj, vm->virtualThreadInspectorCountOffset);
+	if (vthreadInspectorCount < 0) {
+		/* Thread is in transition, wait. */
+		vmFuncs->internalExitVMToJNI(currentThread);
+		VM_AtomicSupport::yieldCPU();
+		/* After wait, the thread may suspend here. */
+		vmFuncs->internalEnterVMFromJNI(currentThread);
+		threadObj = J9_JNI_UNWRAP_REFERENCE(thread);
+		goto retry;
+	} else if (!objectAccessBarrier.inlineMixedObjectCompareAndSwapU64(
+															currentThread,
+															threadObj,
+															vm->virtualThreadInspectorCountOffset,
+															(U_64)vthreadInspectorCount,
+															((U_64)(vthreadInspectorCount + 1)))
+	) {
+		/* Field updated by another thread, try again. */
+		goto retry;
+	}
+}
+
+void
+releaseVThreadInspector(J9VMThread *currentThread, jobject thread)
+{
+	J9JavaVM *vm = currentThread->javaVM;
+	MM_ObjectAccessBarrierAPI objectAccessBarrier = MM_ObjectAccessBarrierAPI(currentThread);
+	j9object_t threadObj = J9_JNI_UNWRAP_REFERENCE(thread);
+	I_64 vthreadInspectorCount = J9OBJECT_I64_LOAD(currentThread, threadObj, vm->virtualThreadInspectorCountOffset);
+	/* vthreadInspectorCount must be greater than 0 before decrement. */
+	Assert_VM_true(vthreadInspectorCount > 0);
+	while (!objectAccessBarrier.inlineMixedObjectCompareAndSwapU64(
+														currentThread,
+														threadObj,
+														vm->virtualThreadInspectorCountOffset,
+														(U_64)vthreadInspectorCount,
+														((U_64)(vthreadInspectorCount - 1)))
+	) {
+		/* Field updated by another thread, try again. */
+		vthreadInspectorCount = J9OBJECT_I64_LOAD(currentThread, threadObj, vm->virtualThreadInspectorCountOffset);
```

Comment:
Should there be a yield here?

Reply:
Same comment on the spin above.

Reply:
This one might not matter as it's a strict decrement operation that can't be interrupted by inspection. A similar argument can be made for the one above.

Reply:
Given the assert that `vthreadInspectorCount > 0` before the loop, this can only fail if another inspector giveup access and decrement the count at sametime. I don't see yield would help much with this type of contention